### ISSUE TITLE
Periodic X diff fails for GFDL data

### ISSUE THREAD

**Issue by jamesp** [xgcm/xgcm#91]
Hi all, I'm trying to collate my analysis routines for the Isca model (fork of GFDL FMS) and think it would make sense to base them on `xgcm` as I can perhaps contribute some back to the core.  For the moment, FMS doesn't seem to conform to CF or Comodo conventions, so I've written a small manifest creator that should put the right metadata in the right place to allow `xgcm.Grid` to read it:
```{python}
def make_isca_grid(ds):
    # make a shell Dataset object that only coordinates
    # so that we don't have to amend the original dataset metadata
    shell = xr.Dataset(coords={k: v.copy() for k, v in ds.coords.items()})
    for coord in shell.coords:
        meta = shell[coord].attrs
        meta['standard_name'] = meta['long_name']
        if 'cartesian_axis' in meta:
            meta['axis'] = meta['cartesian_axis']
            if 'edges' in meta:
                shell.coords[meta['edges']].attrs['c_grid_axis_shift'] = -0.5
                meta['c_grid_axis_shift'] = 0
            elif 'half' in coord:
                meta['c_grid_axis_shift'] = -0.5
    return xgcm.Grid(shell, periodic=['X'])
```
This seems to work well, and returns a Grid object that has the correct discovered C-grid dimensions.

I can then do something like this, which works:
```
g = make_isca_grid(ds)
# ds.ucomp is on (time:10, pfull: 25, lat:64, lon:128)
g.diff(ds.ucomp, 'Z', boundary='fill').mean(('lon', 'time'))

<xarray.DataArray (phalf: 26, lat: 64)>
array([[ 0.073398,  0.194053,  0.348861, ...
```

However, a diff along the periodic direction fails, regardless of whether I set `boundary=` as `None`, `'fill'` or `'extend'`.

```
g.diff(ds.ucomp, 'X', boundary='extend')
...
ValueError: `boundary=extend` is not allowed with periodic axis X.```
```
or 
```
g.diff(ds.ucomp, 'X', boundary=None)
...
ValueError: `boundary` must be `'fill'` or `'extend'````
```

Am I doing something wrong when creating the grid?  Do I need to tell it how the periodic direction works?  I can see reference to `boundary_discontinuity` in the code but it's not obvious to me how I might set this when creating the object.

Notebook with example code above attached, and my example output data is here (approx 8MB): https://www.dropbox.com/s/hdhdr3ebgwyil07/ucomp.nc?dl=0

[xgcm playground.ipynb.zip](https://github.com/xgcm/xgcm/files/1698762/xgcm.playground.ipynb.zip)

**Comment by rabernat** [comment]
Thanks for this bug report! I'll look into it.

FYI I have posted your notebook as a gist here: https://gist.github.com/rabernat/d9253e19b84c13301cb79fddb073679a

**Comment by jbusecke** [comment]
For me this happens when the number of cell faces is unequal to the cell center values. 
Could you post the representation of your custom xgcm.Grid object (e.g. ‘print(g)’)?

The ‘boundary discontinuity’ should only be used when interpolating values which are not periodic (it is used to get values between e.g. longitudes of 0 and 359 as 360 instead of 179.5).

**Comment by rabernat** [comment]
So I just tried your example with the development branch I'm using for #88, and it worked! So that should fix your issue (and also a bunch of others).

Lots of people are waiting on that PR, so this gives me more incentive to just get it merged asap. Hopefully within a day or two.

Thanks again for your interest in xgcm.

**Comment by rabernat** [comment]
> Could you post the representation of your custom xgcm.Grid object (e.g. ‘print(g)’)?

@jbusecke this is all in his example notebook, which I gistified here https://gist.github.com/rabernat/d9253e19b84c13301cb79fddb073679a

**Comment by jamesp** [comment]
Thanks @rabernat and @jbusecke for your very rapid responses! I go off to a seminar and come back to a solution.  I'll fetch #88 and give that a go.

**Comment by jbusecke** [comment]
Edit: I meant to say *continous* instead of periodic above.

Random thought: Why cant I edit posts on the mobile github site...I guess they assume nobody produces typos while typing in the subway ;-)

Also, regarding @rabernat s comment. They dont render notebooks on the mobile site either...grrrr.

**Comment by jamesp** [comment]
@rabernat #88 worked.  Let me know if there is anything I can do to help test/document and get it into master sooner.

@jbusecke  Thanks for the clarification on `boundary_discontinuity`.  And there's definitely a gap in the market out there for a good mobile interactive notebook viewer! :)

**Comment by jamesp** [comment]
On a related, but off topic note, I have some xarray functions for calculating metrics e.g. overturning streamfunction, brunt vaisala frequency, from Isca model output.  Also an fft routine, inspired in part by something @rabernat posted in another issues thread, that can also be used as a dataarray accessor i.e. `ds.vorticity.fft(('time', 'lon'))` (https://github.com/jamesp/isca_xarray/blob/master/iscaxr/analysis/spectral.py#L24).

Are these in scope for the xgcm project?  If so I can look at generalising them to your grid specification and roll into this repo; if not I'll leave them as analysis specific to the Isca model.

**Comment by jbusecke** [comment]
I think, particular for spectral analysis https://github.com/rabernat/xrft would be the package to contribute (not directly dependent on the xgcm 'magic'). I am very interested in building vector calculus (like a vorticity operator above) into xgcm, in a way that automatically recognizes/rebuilds appropriate metrics (distances, area, volume), needed for the calculation. This should be flexible enough to support different grid types and different methods (e.g. finite volume vs finite difference). There should be issues open about those topics, but I cant find them right now (on the train again :-)). We could also start a module based on xgcm and perhaps later integrate it? Cheers

**Comment by rabernat** [comment]
closed by #88
